exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1858 @@
+		)
+	}
+	for x := range assets {
+		pairs, err := ok.GetEnabledPairs(assets[x])
+		if err != nil {
+			return nil, err
+		}
+		for y := range defaultSubscribedChannels {
+			if defaultSubscribedChannels[y] == OkxChannelCandle5m ||
+				defaultSubscribedChannels[y] == OkxChannelTickers ||
+				defaultSubscribedChannels[y] == OkxChannelOrders ||
+				defaultSubscribedChannels[y] == OkxChannelOrderBooks ||
+				defaultSubscribedChannels[y] == OkxChannelOrderBooks5 ||
+				defaultSubscribedChannels[y] == OkxChannelOrderBooks50TBT ||
+				defaultSubscribedChannels[y] == OkxChannelOrderBooksTBT ||
+				defaultSubscribedChannels[y] == OkxChannelTrades {
+				for p := range pairs {
+					subscriptions = append(subscriptions, stream.ChannelSubscription{
+						Channel:  defaultSubscribedChannels[y],
+						Asset:    assets[x],
+						Currency: pairs[p],
+					})
+				}
+			} else {
+				subscriptions = append(subscriptions, stream.ChannelSubscription{
+					Channel: defaultSubscribedChannels[y],
+				})
+			}
+		}
+	}
+	return subscriptions, nil
+}
+
+// wsProcessPushData processes push data coming through the websocket channel
+func (ok *Okx) wsProcessPushData(data []byte, resp interface{}) error {
+	if er := json.Unmarshal(data, resp); er != nil {
+		return er
+	}
+	ok.Websocket.DataHandler <- resp
+	return nil
+}
+
+// Websocket Trade methods
+
+// WSPlaceOrder places an order thought the websocket connection stream, and returns a SubmitResponse and error message.
+func (ok *Okx) WSPlaceOrder(arg *PlaceOrderRequestParam) (*PlaceOrderResponse, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.InstrumentID == "" {
+		return nil, errMissingInstrumentID
+	}
+	if !(strings.EqualFold(TradeModeCross, arg.TradeMode) || strings.EqualFold(TradeModeIsolated, arg.TradeMode) || strings.EqualFold(TradeModeCash, arg.TradeMode)) {
+		return nil, errInvalidTradeModeValue
+	}
+	if !(strings.EqualFold(arg.Side, "buy") || strings.EqualFold(arg.Side, "sell")) {
+		return nil, errMissingOrderSide
+	}
+	if !(strings.EqualFold(arg.OrderType, "market") || strings.EqualFold(arg.OrderType, "limit") || strings.EqualFold(arg.OrderType, "post_only") ||
+		strings.EqualFold(arg.OrderType, "fok") || strings.EqualFold(arg.OrderType, "ioc") || strings.EqualFold(arg.OrderType, "optimal_limit_ioc")) {
+		return nil, errInvalidOrderType
+	}
+	if arg.QuantityToBuyOrSell <= 0 {
+		return nil, errInvalidQuantityToButOrSell
+	}
+	if arg.OrderPrice <= 0 && (strings.EqualFold(arg.OrderType, "limit") || strings.EqualFold(arg.OrderType, "post_only") ||
+		strings.EqualFold(arg.OrderType, "fok") || strings.EqualFold(arg.OrderType, "ioc")) {
+		return nil, fmt.Errorf("invalid order price for %s order types", arg.OrderType)
+	}
+	if !(strings.EqualFold(arg.QuantityType, "base_ccy") || strings.EqualFold(arg.QuantityType, "quote_ccy")) {
+		arg.QuantityType = ""
+	}
+	randomID := common.GenerateRandomString(32, common.SmallLetters, common.CapitalLetters, common.NumberCharacters)
```

Comment:
We already have this kind of functionality in the form of:
```
		No:        z.Websocket.Conn.GenerateMessageID(true),
	}

	request.Sign, err = z.wsGenerateSignature(creds.Secret, request)
	if err != nil {
		return nil, err
	}

	resp, err := z.Websocket.Conn.SendMessageReturnResponse(request.No, request)
```

This allows the websocket to track responses without needing this extra random string function you've written. Checkout zb_websocket for how its done there.

Reply:
but, here I thought It would be fine to use a combination of characters(upper and lower case) and numbers.
Even though I didn't know this _websocket.Conn.GenerateMessageID(true),_ function.